### ISSUE TITLE
Python API: Algebraic Type Check

### ISSUE THREAD

**Issue by bdrodes** [CVC4/CVC4#4961]
What's the python API to perform a type check on an algebraic type, see the `is_String` assertion below. 
```
OPTION "dt-nested-rec" TRUE;
OPTION "strings-exp" TRUE;
OPTION "produce-models" TRUE;

DATATYPE
    MyType =
      Int(int: INT)
    | Bool(bool: BOOLEAN)
    | String(string: STRING)
END;

x : MyType; 

ASSERT is_String(x);
```

The following python code, I define x, and while I can print x and see `(String "test")`, I don't see a way to pull that type information out directly. 
```
import pycvc4
slv = pycvc4.Solver()

slv.setOption("produce-models", "true")
slv.setOption("strings-exp", "true")
slv.setOption("incremental", "true")


my_int = slv.mkDatatypeConstructorDecl('Int')
my_int.addSelector('int', slv.getIntegerSort())
my_bool = slv.mkDatatypeConstructorDecl('Bool')
my_bool.addSelector('bool', slv.getBooleanSort())
my_string= slv.mkDatatypeConstructorDecl("String")
my_string.addSelector('string', slv.getStringSort())

my_sort = slv.declareDatatype('MyType', [my_int, my_bool, my_string])
my_type = my_sort.getDatatype()

x = slv.mkTerm(pycvc4.kinds.ApplyConstructor, my_type.getConstructorTerm('String'), slv.mkString("test"))
```

**Comment by 4tXJ7f** [comment]
There are two aspects to this: using a tester in an assertion and pulling out the constructor from a term that represents a constructor application.

When you use:

```
ASSERT is_String(x);
```

Then you are asserting that `x` is a datatype constructed with the constructor `String`. With the Python API, the equivalent would be:

```
x = slv.mkConst(my_sort, 'x')                                                                
a = slv.mkTerm(pycvc4.kinds.ApplyTester, my_type.getConstructor('String').getTesterTerm(), x)
                                                                                             
# ((_ is String) x)                                                                          
print(a)                                                                                     
                                                                                             
slv.assertFormula(a)                                                                         
```

When you write:
```
x = slv.mkTerm(pycvc4.kinds.ApplyConstructor, my_type.getConstructorTerm('String'), slv.mkString("test"))
```

Then you are constructing a term that represents the application of a constructor to the string constant `"test"` but this is just a regular term and not a special constant.

If you want to inspect such a term (e.g. because you got it from `getValue()`), you can do:

```
x = slv.mkConst(my_sort, 'x')                                                                
a = slv.mkTerm(pycvc4.kinds.ApplyTester, my_type.getConstructor('String').getTesterTerm(), x)   
slv.assertFormula(a)                    
slv.checkSat()                          
                                        
vx = slv.getValue(x)                    
                                        
# (String "")                           
print(vx)                               
print(vx.getKind())                     
                                        
vxc = [c for c in vx]                   
                                        
# String                                
print(vxc[0])                           
```

(Note: `vxc = [c for c in vx]` is a workaround for the fact that we seem to be missing `__getitem__` in our `Term` class right now, I will fix this)

Hope this helps.

**Comment by bdrodes** [comment]
Thanks. I think this gets me what I need. Is there a ticket I can follow for the __getitem__ fix? I assume the fix will let me perform introspection without having to go through a getValue? Your workaround above at least works that way. 


Also, speaking of __getitem__ issues. I had to modify __getitem__ in the cvc4.pxi Datatype class. I've seen examples where I should be able to do `my_type['String']`, but I was getting 

```
cvc4.pxi in pycvc4.Datatype.__getitem__()
NameError: name 'name' is not defined
```
Modified as follows:

```
         elif isinstance(index, str):
-            dc.cdc = self.cd[(<const string &> name.encode())]
+            dc.cdc = self.cd[(<const string &> index.encode())]
```

**Comment by 4tXJ7f** [comment]
The support for `__getitem__` in `Term` has been merged now along with the fix for `__getitem__` in `Datatype` (#4998). Now, you can omit `vxc = [c for c in vx]` and instead directly do `vx[0]`.

`getValue()` is actually independent of that. I just wanted to pick an example where you get a `Term` from the solver and you want to figure out what constructor is being used. Both `getValue()` and `mkTerm()` return the same kind of `Term`s (in this case, a term of kind `ApplyConstructor` with two children, the constructor and the argument to the constructor). Another reason for picking that example was to make a clear distinction between `ASSERT is_String(x);` and analyzing the structure of a term.

I hope this makes sense.

**Comment by bdrodes** [comment]
Thanks. I think this resolves my original question on this ticket, but I opened another ticket (https://github.com/CVC4/CVC4/issues/5014) to address a related concern of if I can retrieve constant values from either algebraic types or any base type. Two things I wanted to achieve were the ability to grab constructors and to grab constant values, but it seems getting constant values is not straightforward even for ordinary types. See the new issue for details.
